### DISCUSSION TITLE
How to use led-mapper with effects written using the traditional XY() mapping function

### DISCUSSION THREAD

**Discussion by moose4lord** [jasoncoon/led-mapper#4] [Q&A]
Hello, I was searching the internet for a tool to map FastLED effects to my cube-shaped WS2812B project and your led-mapper looks great. Unfortunately, I'm confused about how to use it with effects that have been written using the "traditional" method of providing an XY() function to map a matrix of LEDs (like the FastLED XYMatrix, SmartMatrix and some of the Noise example sketches). 

My cube is made of 5 8x8 WS2812B panels, with the 320 LEDs laid out in led-matrix as a 24x24 matrix like this:

```
         -------
        | BACK  |
 -----------------------
| LEFT  |  TOP  | RIGHT |
 -----------------------
        | FRONT |
         -------
```

Since led-mapper throws away info about missing LEDs in the matrix, it seems that it couldn't be used for XY() style mapping. Or am I missing something?

Thanks

**Comment by jasoncoon** [comment]
Hey, it's not terribly efficient, but you could create an XY function, equivalent to the [XYMatrix example](https://github.com/FastLED/FastLED/blob/master/examples/XYMatrix/XYMatrix.ino#L88), like this:

```c
uint16_t XY(uint8_t x, uint8_t y) {
   for (uint16_t i = 0; i < NUM_LEDS; i++) {
      if (coordsX[i] == x && coordsY[i] == y)
         return i;
   }
   return -1;
}
```

Alternatively, you could create the XY function in any other way you see fit, including [Garrett Mace's mapper](https://macetech.github.io/FastLED-XY-Map-Generator) and use multiple different mapping methods in the same sketch.

You can see how I mapped the surface of a cube with six 8x8 panels here: https://github.com/jasoncoon/esp8266-fastled-webserver/blob/ultim8x8-cube/Map.h#L1

[ULTiM8x8 Cube + FastLED + ESP8266 Web Server](https://github.com/jasoncoon/esp8266-fastled-webserver/blob/ultim8x8-cube/README.md)

Good luck! Can't wait to see what you come up with!

**Reply by moose4lord** [reply]
Thanks for getting back to me so quickly Jason. I see what you mean about the nested loop solution. Although XY() functions typically expect X and Y values to have a range of 0 to (MAX_X - 1), like this:
```
  for( uint8_t y = 0; y < kMatrixHeight; y++) { 
    for( uint8_t x = 0; x < kMatrixWidth; x++) {
      leds[ XY(x, y)]  = [some color generating function];
    }
  }
```
and the coordsX and coordsY arrays created by led_mapper are normalized for matrix dimensions that range from 0 to 255, so it would be a little more complicated than that I think. And also quite inefficient, as you say.

I scrolled to the bottom of the led-mapper web page and saw the "Parsed Layout" array and that looks like it will work. I changed the nulls to -1 and the syntax to create C nested arrays, so the YX() function is a simple table lookup. It does add an extra 1152 bytes to the sketch, but I'm running on an ESP32, so memory isn't an issue. With that added, the XYMatrix example sketch is working great. I guess I should have scrolled to the bottom of the page a little sooner. :)

While I was perusing the ULTiM8x8 Cube project you linked to, I noticed you made some PCB boards to make interfacing ESP8266's to RGB LEDs easier. Dang, I wish I would have known that a year ago. My ESP8266 projects would have looked much more professional. Actually I'm tempted to get one for my ESP32 project, since my [Mini ESP32](https://www.aliexpress.us/item/3256801459857540.html?spm=a2g0o.store_pc_allProduct.8148356.3.539961bcu2WqCq&pdp_npi=2%40dis%21USD%21US%20%243.90%21US%20%243.43%21%21%21%21%21%40210318be16658403115506537e765e%2112000027029908865%21sh&gatewayAdapt=glo2usa&_randl_shipto=US) is supposed to be Wemos D1 Mini compatible. Still I would like my design to incorporate a tilt switch, so I can wake the ESP32 by shaking the cube, as well as a MOSFET so I can completely disconnect power from the WS2812Bs and reduce power when the ESP32 is sleeping. So if you're ever thinking of creating a PCB v3.0 design, those would be nice. Just sayin'. :)

**Reply by moose4lord** [reply]
Oops, meant to attach a video of what I have so far....

https://user-images.githubusercontent.com/15672113/195990253-532e6f9d-d728-4b6c-a7ed-6fa7113aac1c.mp4

**Reply by jasoncoon** [reply]
Nice!

I no longer sell any shields, but I've shared them all on OSH Park if you still want any: https://oshpark.com/profiles/jasoncoon

**Comment by wile1411** [comment]
Hey Jason - Just a thought for those using multiple methods of XY lookups as well as LED offset that your tool does so well.
Could the tool also generate an XY lookup table output (with optional -1 for any non-existing LEDS) as a third output alongside the two existing options?
